src/msssim.jl

```diff
@@ -0,0 +1,84 @@
+    RGB first.
+```
+
+# Reference
+
+[1] Wang, Z., Bovik, A. C., Sheikh, H. R., & Simoncelli, E. P. (2004). Image quality assessment: from error visibility to structural similarity. _IEEE transactions on image processing_, 13(4), 600-612.
+
+"""
+struct MSSSIM{N} <: FullReferenceIQI
+    kernel::AbstractArray{<:Real}
+    W::NTuple{N, <:Tuple} # number of scales inferred from length
+    function MSSSIM(kernel, W)
+        ndims(kernel) == 1 || throw(ArgumentError("only 1-d kernel is valid"))
+        issymetric(kernel) || @warn "MSSSIM kernel is assumed to be symmetric"
+        all(length.(W) .== 3) || throw(ArgumentError("(α, β, γ) required for all scales, instead it's $(W)"))
+        all([ all(x .>= 0) for x in W ]) || throw(ArgumentError("α, β, γ should be non-negative for all scales, instead it's $(W)"))
+        new{length(W)}(kernel, W)
+    end
+end
+
+# Weights for α, β, γ in [1]
+const MSSSIM_W = (0.0448, 0.2856, 0.3001, 0.2363, 0.1333) # α, β, γ for scales 1 through 5
+
+MSSSIM(kernel, W::NTuple{N, <:Real}) where N = MSSSIM(kernel, map(x->(x, x, x), W)) # shorthand for αᵢ=βᵢ=γᵢ for all scales i
+
+# USE SSIM_KERNEL as default kernel
+MSSSIM(kernel=SSIM_KERNEL) = MSSSIM(kernel, MSSSIM_W)
+
+(iqi::MSSSIM)(x, ref) = _msssim_map(iqi, x, ref)
+assess_msssim(x, ref) = MSSSIM()(x, ref)
+
+const DOWNSAMPLE_FILTER = ones(2, 2)./4 # as per author's implementaion
+
+# SSIM does not allow for user specifying peakval and K, so we don't allow it here either
+function _msssim_map(iqi::MSSSIM, x::GenericImage, ref::GenericImage)
+    if size(x) ≠ size(ref)
+        err = ArgumentError("images should be the same size, instead they're $(size(x))-$(size(ref))")
+        throw(err)
+    end
+
+    level = length(iqi.W) # number of scales
+    T = promote_type(float(eltype(ref)), float(eltype(x)))
+    x = of_eltype(T, x)
+    ref = of_eltype(T, ref)
+
+    # check if no. of levels are >= 1
+    level < 1 && throw(ArgumentError("MS-SSIM needs at least one scale"))
+
+    # check if weights are valid - as per authors implementaion
+    sum(sum.(iqi.W)) == 0 && throw(ArgumentError("MS-SSIM must have at least one weight > 0"))
+
+    # downsampling window
+    window = kernelfactors(Tuple(repeated(DOWNSAMPLE_FILTER, ndims(ref))))
+
+    mean_cs = []
+    for i in 1:level-1
+        cs = SSIM(iqi.kernel, (zero(typeof(iqi.W[i][1])), iqi.W[i][2], iqi.W[i][3]))(x, ref)
+        append!(mean_cs, cs)
+
+        imfilter(x, window, "symmetric")
+        imfilter(ref, window, "symmetric")
+
+        # downsampling
+        x = x[ntuple(i->first(axes(x, i)) : 2 : last(axes(x, i)), ndims(x))...]
+        ref = ref[ntuple(i->first(axes(ref, i)) : 2 : last(axes(ref, i)), ndims(ref))...]
+    end
+
+    # last scale
+    lcs = SSIM(iqi.kernel, (iqi.W[end][1], iqi.W[end][2], iqi.W[end][3]))(x, ref)
+    append!(mean_cs, lcs)
+
+    return min(prod(mean_cs), 1.0)
+
+    # TODO: Add sum option as well from author's implementaion
```

Comment:
The author's implementation gives two options for collecting the results of the several SSIM runs, `sum` and `product`. From the MATLAB implementation:
```(MATLAB)
if (method == 'product')
   overall_mssim = prod(mcs_array(1:level-1).^weight(1:level-1))*(mssim_array(level).^weight(level));
else
   weight = weight./sum(weight);
   overall_mssim = sum(mcs_array(1:level-1).*weight(1:level-1)) + mssim_array(level).*weight(level);
```
This is nowhere mentioned in the paper though, and does not appear in the tensorflow implementation either, not sure if it needs to be added.

Reply:
Not knowing how other implementations do, I'll take whatever choice you make.

For me, looks like it's only a matter of `log`, since `log(product_method_result) == sum_method_result`?

I've met similar situations when I implemented SSIM, see https://github.com/JuliaImages/ImageQualityIndexes.jl/pull/3#discussion_r289122499. If there's no standard between the implementations, choose the one you think is the best, and if necessary, document these differences.

Reply:
Yes, the powers get multiplied and products become sums, but it doesn't seem to actually compute the `log` of the values.
I will let this one pass for now, because there is no indication as to why this is useful, and tensorflow doesn't implement it either. It will unnecessarily confuse users in my opinion.